Day-2/Theory/README.md

```diff
@@ -0,0 +1,369 @@
+    import React from 'react';
+
+    class App extends React.Component {
+        constructor(props) {
+            // we will see later what super(props) means
+            super(props);
+
+            // initialize the state
+            this.state = {
+                x: "This is x from state",
+                y: "This is y from state"
+            }
+        }
+
+        render() {
+            return (
+                <div>
+                    <div>{this.state.x}</div>
+                    <div>{this.state.y}</div>
+                </div>
+            )
+        }
+    }
+    ```
+
+- we can update the state in our component using *setState* method:
+
+    ```JavaScript
+    this.setState({x: "This is the new x"})
+    ```
+
+- *setState* is asynchronous, so be careful if you call some pieces of code in the same update cycle:
+
+    ```JavaScript
+    // wrong way
+    this.setState({value: this.state.value + 1})
+    ```
+
+    ```JavaScript
+    // right way
+    this.setState(({value}) => ({value: value + 1}))
+    ```
+
+## Props
+
+- props are similar to method arguments
+- they are passed to a child component
+- props are immutable
+
+    ```JavaScript
+    import React from 'react';
+
+    class Dog extends React.Component {
+        render() {
+            return (
+                <div>
+                    <label>This is a {this.props.color} dog.</label>
+                </div>
+            )
+        }
+    }
+
+    export default Dog;
+
+    const dog = <Dog color="white"></Dog>
+    ```
+
+## Components
+
+- components allow us to split de UI into smaller, independent and reusable pieces
+- there are two categories of components:
+  - class components
+  - function components
+
+### Class Components
+
+- the name should start with upper case letter
+- we should include *extends React.Component* statement to give our component access to some functions
+- need a *render()* method which returns HTML
+- a class component has its own state, so can be also called *stateful* component
+- e.g.:
+
+  ```JavaScript
+    class Student extends React.Component {
+        state = {
+            name: "Andrei",
+            age: 22,
+            class: "A1",
+        }
+
+        changeClass = () => {
+            this.setState({class: "A2"});
+        }
+
+        render() {
+            return (
+                <div>
+                    <div>{this.state.name}</div>
+                    <button onClick={this.changeClass}>Change Class</button>
+                </div>
+            )
+        }
+    }
+    ```
+
+- to use this component into our application, it will be called through a HTML syntax: *<Student\/>*
+
+  ```JavaScript
+  ReactDOM.render(<Student />, document.getElementById('root'));
+  ```
+
+### The Lifecycle of a Component
+
+- each component has several *lifecycle methods* that we can override to execute code at a particular time
+- the most common ones are:
+![Common-lifecycle-methods](img/common_lifecycle_methods.png)
```

Comment:
Please fix the url and add source: source: http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/

Reply:
Done. Thanks! :)